### PR TITLE
Adding unsupported questions types

### PR DESCRIPTION
**Pull request by dariosalviwork**
this would fix issues: #228 , #229 , #216 , #159

### CODE REVIEW

**Comment by nrlakin** [comment]
Close/reopen to force travis-ci build.

**Comment by nrlakin** [comment]
Travis CI has a bug--there's no .travis.yml in your branch because you forked the repo before it existed, so it's failing to build for the unit tests. Mind updating the pull request by checking out the corresponding local branch, pulling from the upstream master branch, then pushing back to your development branch? That will add the Travis config file to the pull request so it can run the unit tests. There are no merge conflicts, so it should be smooth. Sorry for the hassle, shouldn't be an issue going forward as new pull requests will have .travis.yml already.

File: skin/src/main/java/org/researchstack/skin/task/SmartSurveyTask.java [hunk 1]
```diff
@@ -116,8 +122,13 @@ else if(type.equals("DateConstraints"))
         {
             answerFormat = new DateAnswerFormat(AnswerFormat.DateAnswerStyle.Date);
         }
+        else if(type.equals("DurationConstraints"))
+        {
+            answerFormat = new DurationAnswerFormat(constraints.step, constraints.durationUnit);
+        }
         else
         {
+            LogExt.e(SmartSurveyTask.class, "Survey question has answer type not supported:" + type);
```
Comment (mikecarroll) [inline skin/src/main/java/org/researchstack/skin/task/SmartSurveyTask.java]: This is useful -- thanks for adding it! But how do you feel about going the "stronger" route and actually raising an exception here, or at least a warning? I feel like it's better to catch unsupported question types here with an explicit exception than let this log message get buried in the stack trace somewhere. Thoughts? @dariosalviwork @nrlakin @crayne

File: backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.researchstack.backbone.answerformat.DurationAnswerFormat;
+import org.researchstack.backbone.result.StepResult;
+import org.researchstack.backbone.step.QuestionStep;
+import org.researchstack.backbone.step.Step;
+import org.researchstack.backbone.utils.TextUtils;
+import org.researchstack.backbone.utils.ViewUtils;
+
+/**
+ * Implementation of the question type "duration"
+ * the JSON can specify a unit and a step, here they are not used, and probably not either in ResearchKit
```
Comment (mikecarroll) [inline backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java]: Not sure what you mean in the comments here. Do you mind explaining? Just want to make sure the documentation is easy for others to follow.

File: backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ */
+public class DurationQuestionBody implements StepBody
+{
+    //-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*
+    // Constructor Fields
+    //-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*
+    private QuestionStep        step;
+    //I'm assuming the result is given as the total number of minutes
```
Comment (mikecarroll) [inline backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java]: I think you can change to simply say "The result should be given as the total number of minutes." and move it to the large comment block before the method declaration, above.

**Comment by mikecarroll** [comment]
This is excellent work, @dariosalviwork! Just a few points for clarification/discussion (besides the hassle of rebasing to handle the Travis CI stuff -- sorry about that).

Also, apologies for the delay in reviewing: we wanted to get Travis up and running before merging more PRs, just to make sure we weren't accidentally introducing regressions. If you have any tests for these new answer types, would love it if you could commit them, too.

File: skin/src/main/java/org/researchstack/skin/task/SmartSurveyTask.java (hunk 1 above)
Comment (dariosalviwork) [inline skin/src/main/java/org/researchstack/skin/task/SmartSurveyTask.java]: we can throw a RuntimeException, but then why is there that UnknownAnswerFormat at all?
We would need to ask the people who designed this...

File: backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java (hunk 2 above)
Comment (dariosalviwork) [inline backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java]: I've updated the text, see my new pull request

File: backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java (hunk 3 above)
Comment (dariosalviwork) [inline backbone/src/main/java/org/researchstack/backbone/ui/step/body/DurationQuestionBody.java]: OK done

**Comment by dariosalviwork** [comment]
sorry for being absent for so long, but I've been very busy with other commitments!
shall we try again now?

**Comment by suomi35** [comment]
Can we go ahead and merge these changes?

**Comment by mikecarroll** [comment]
Yes, sorry, merging very soon. Just waiting for the results from a security audit of the whole platform (and any fixes that might come from that) before proceeding with PRs. Should be any day now -- sorry for the delay.

**Comment by dariosalviwork** [comment]
Hi people, I have fixed some issues here.

**Comment by dariosalviwork** [comment]
Hi again, I have added an extension to be able to observe the activity lifecycle from a Task. It allows Tasks to control the view dynamically. Very much needed in the case of a fitness test for example.

**Pull request merged**